C/primitive/elements/primitiveJetNode.inc

```diff
@@ -2,250 +2,250 @@
-, .cost = 165 /* milli weight units */
+, .cost = 79 /* milli weight units */
 }
 ,[ALL_8] =
 { .tag = JET
 , .jet = all_8
-, .cmr = {{0xdc098854u, 0xdd06ae1eu, 0x6e3e73a4u, 0xae94d0b2u, 0xacce5cb3u, 0xecc12e8cu, 0xb8167f7bu, 0x6eaa4069u}}
+, .cmr = {{0x0d3cbfc2u, 0xcc2a9a06u, 0x28a2ff8au, 0xfc0a5f93u, 0x173ede4cu, 0x234c5721u, 0x4c5d74afu, 0x6b873ce9u}}
 , .sourceIx = ty_w8
 , .targetIx = ty_b
-, .cost = 113 /* milli weight units */
+, .cost = 76 /* milli weight units */
 }
 ,[AND_1] =
 { .tag = JET
 , .jet = and_1
-, .cmr = {{0x2dc2dbccu, 0x69c12c78u, 0x30df1170u, 0xd9e93a35u, 0xd8284cc8u, 0x15916aebu, 0x3b1b95efu, 0xdaa92c26u}}
+, .cmr = {{0x56d1d4d6u, 0xc1e11f60u, 0xd29660b1u, 0x81a1335au, 0x4f511d3bu, 0x9c04a333u, 0xe9e01418u, 0xb0d1818fu}}
 , .sourceIx = ty_w2
 , .targetIx = ty_b
-, .cost = 159 /* milli weight units */
+, .cost = 79 /* milli weight units */
 }
 ,[AND_16] =
 { .tag = JET
 , .jet = and_16
-, .cmr = {{0x4f42c688u, 0xcc3ceb76u, 0x1f662a0bu, 0x8e77b196u, 0x42dab540u, 0x7d9a873cu, 0xd9c18609u, 0x977ebd91u}}
+, .cmr = {{0xd77a1144u, 0xc52ed2b7u, 0x6988df9bu, 0x29cd5397u, 0x8c44c6d5u, 0x40f4bc54u, 0x56d0b3b5u, 0x3cd400e5u}}
 , .sourceIx = ty_w32
 , .targetIx = ty_w16
-, .cost = 195 /* milli weight units */
+, .cost = 88 /* milli weight units */
 }
 ,[AND_32] =
 { .tag = JET
 , .jet = and_32
-, .cmr = {{0x1680cfd2u, 0x643ad389u, 0x6bfb45e9u, 0x6fc62dc2u, 0xbb0cc0acu, 0x486fa833u, 0x30a8890cu, 0x09ee969bu}}
+, .cmr = {{0x4382f934u, 0x3379d442u, 0x2e2a0151u, 0x04cf55dcu, 0x54a8f71eu, 0xefb06229u, 0x26d64b6cu, 0xe22bea11u}}
 , .sourceIx = ty_w64
 , .targetIx = ty_w32
-, .cost = 175 /* milli weight units */
+, .cost = 94 /* milli weight units */
 }
 ,[AND_64] =
 { .tag = JET
 , .jet = and_64
-, .cmr = {{0xa2dd3d49u, 0xf12cd442u, 0xcea7c832u, 0x3868be3cu, 0xe6e9af67u, 0x6f725110u, 0xf3ce494eu, 0xdd6a5a4au}}
+, .cmr = {{0x3eb8b42du, 0x815f0fc3u, 0x9a3499b0u, 0x354efd05u, 0xa603f5efu, 0x5e511766u, 0xc0112cc1u, 0x48e21155u}}
 , .sourceIx = ty_w128
 , .targetIx = ty_w64
-, .cost = 221 /* milli weight units */
+, .cost = 93 /* milli weight units */
 }
 ,[AND_8] =
 { .tag = JET
 , .jet = and_8
-, .cmr = {{0x27554178u, 0x198658f5u, 0xbfb5b041u, 0x30898742u, 0xbf82db60u, 0x39dcbb17u, 0x266dfe03u, 0x0cc12075u}}
+, .cmr = {{0x0e0ee07cu, 0x749a3fa7u, 0x1296b9d4u, 0x59b06a0au, 0x4bffb7f5u, 0xbb99a925u, 0x4a7d13a8u, 0xce444cd7u}}
 , .sourceIx = ty_w16
 , .targetIx = ty_w8
-, .cost = 159 /* milli weight units */
+, .cost = 91 /* milli weight units */
 }
 ,[ANNEX_HASH] =
 { .tag = JET
 , .jet = annex_hash
-, .cmr = {{0xca1beda0u, 0x4f1fe8cbu, 0xb08cd69fu, 0x3959d442u, 0xfe5ab015u, 0xc6923e17u, 0x1d8d0686u, 0xef31488fu}}
+, .cmr = {{0x07eebc33u, 0xfde1ad63u, 0xef136901u, 0xbaaea6fdu, 0xe7228d7cu, 0x60d179b9u, 0x47bd5de5u, 0xe07ab785u}}
 , .sourceIx = ty_pppmw256pmw128pmw64pmw32pmw16mw8pw64w256mw256
 , .targetIx = ty_ppmw256pmw128pmw64pmw32pmw16mw8pw64w256
-, .cost = 3065 /* milli weight units */
+, .cost = 241 /* milli weight units */
```

Comment:
There has been a substantial drop in costs for `AnnexHash` and similar function such as `NonceHash` and `OutpointHash` and `AssetAmountHash`.

Reply:
@apoelstra So long as you are happy with the state of your distributions for Sha256 contexts, then I'm find with this.  We did speed up hashing somewhat, and maybe the benchmark used to have extra overhead marshaling the complex context type.

Reply:
These methods are all transaction context methods that don't actually do any hashing themselves, just lookups.

I did mess with the transaction environment but I don't believe that I meaningfully changed the distribution.

So I'm happy with this.

Reply:
For the tx environment jets I didn't change any of the marshalling logic. But the new numbers still look right and the old ones look wrong. Let me spend a few minutes trying to bisect whether there is a specific change I made that caused this drop.

And indeed, when I run these benchmarks locally I see roughly the same results for asset_amount_hash on `master` or on my `benchmark` branch or immediately before/after my change to the tx environment. Same results before the recent update to libsimplicity, and with the libsimplicity from Aug 2023 when we did the last benchmark update.

My suspicion is that because these benchmarks are run near the end of the list of jets, and Sanket works late while I work early, is that the old numbers were artificially slowed down by my waking up near the end of his run and using the benchmark machine for other stuff.

Reply:
I'm not so concerned that this is a transaction environment issue, after all there are several other jets that use the transaction environment.

What makes these jets special is that they take a `Sha256Ctx` input.  And maybe if you are only producing trivial Sha256 contexts, the numbers could end up incorrectly too small.

Reply:
The code I'm using is identical to the code we used last time, and the results I get are identical no matter what old versions of rust-simplicity I run them on. So I don't know what to make of this.

Reply:
As one last test, do you want to try benchmarking with `-D NO_SHA_NI_FLAG`?

Reply:
Sure. I will start that overnight tonight.

Reply:
Hmm, so, actually, in my Rust build neither '__SHA__` nor __SSH4_1__` are defined, so none of these benchmarks are using the accellerated sha at all.

Reply:
If I force these symbols to be set I get a pile of errors like 

```
cargo:warning=In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/14.1.1/include/immintrin.h:39,
  cargo:warning=                 from depend/simplicity/sha256_x86.inc:5,
  cargo:warning=                 from depend/simplicity/sha256.c:208:
  cargo:warning=/usr/lib/gcc/x86_64-pc-linux-gnu/14.1.1/include/smmintrin.h: In function ‘_mm_testz_si128’:
  cargo:warning=/usr/lib/gcc/x86_64-pc-linux-gnu/14.1.1/include/smmintrin.h:69:10: error: implicit declaration of function ‘__builtin_ia32_ptestz128’; did you mean ‘__builtin_ia32_pabsd128’? [-Wimplicit-function-declaration]
  cargo:warning=   69 |   return __builtin_ia32_ptestz128 ((__v2di)__M, (__v2di)__V);
  cargo:warning=      |          ^~~~~~~~~~~~~~~~~~~~~~~~
  cargo:warning=      |          __builtin_ia32_pabsd128
  cargo:warning=/usr/lib/gcc/x86_64-pc-linux-gnu/14.1.1/include/smmintrin.h: In function ‘_mm_testc_si128’:
```

so this is something messed up with my compiler toolchain, not something "simple" like cc-rs failing to define certain flags.

Reply:
It's probably for the best that the benchmarks are not using accelerated SHA.

Reply:
I tried doing this and the code compiled (and by adding `#error` in various places I confirmed we were actually compiling the contents of the .inc file) but the sha2 benchmarks didn't change.

I'm getting a bit tired of poking at these benchmarks but if there's something simple you think I'm missing I'd like to give it a shot.

Reply:
Do you have a PR or otherwise have a viewable source of the benchmark generating code?

Reply:
Yes, I have https://github.com/BlockstreamResearch/rust-simplicity/pull/239

I have the following diff on top of that

```diff
diff --git a/simplicity-sys/build.rs b/simplicity-sys/build.rs
index 6567e8e..223db13 100644
--- a/simplicity-sys/build.rs
+++ b/simplicity-sys/build.rs
@@ -24,6 +24,12 @@ fn main() {
     let mut build = cc::Build::new();
     build
         .std("c11")
+//        .define("NO_SHA_NI_FLAG", Some("1"))
+//        .define("__SHA__", Some("1"))
+//        .define("__SSE4_1__", Some("1"))
+//        .define("__x86_64__", Some("1"))
+        .flag_if_supported("-msse4")
+        .flag_if_supported("-msha")
         .flag_if_supported("-fno-inline-functions")
         .opt_level(2)
         .files(jet_files)
```